Bind services to clusters

bandesz opened the pull request:
## What

We had a discussion about the future of services and we think we've come up with a solution which will simplify things in the future and give us a more constrained scope:
 * all services will be owned by a single cluster (which means the cluster field will be required)
 * there will be an option to share a service with other clusters (which will only allow to manage service access (service credentials in an other cluster) - not part of this story
 * there will be two types of services like now: cloud services and application services (in-cluster services). One technical distinction is that application services will need a cluster namespace, while cloud services won't.

Technically an application service is where `kore.appvia.io/platform=Kubernetes`, everything else is a cloud service.

### Changes

 * New cluster page layout with tabs
 * The cloud services were moved to the cluster page
 * Tab permalinks on the cluster page (thanks to @davet1985)
 * The service page was moved under the cluster URL (thanks to @davet1985)
 * Add a ServiceAccessEnabled flag to service kinds (if false, you can't create service accesss)
 * Add a ServiceAccessDisabled flag to service plans (if true, you can't create service access)
 * Strict handling of the feature gates everywhere
    * the services menu on the left should only show up when the services feature gate is enabled
    * the services config page should not show app service kinds if the feature gate is not enabled.
 * The Dummy service will create a service kind for AWS, GCP, Azure and Kubernetes for easier testing
 * Add cluster and namespace options to `kore create service`
 * various smaller fixes/improvements (see commits)

### New cluster page layout

![Screenshot 2020-06-09 at 18 17 05](https://user-images.githubusercontent.com/868104/84189204-ce35d800-aa8c-11ea-9966-fbd4a5efee06.png)

### Testing

1. Make sure the `application_services` feature gate is enabled on the UI. If you use `make run` or `make run-prod`, it is by default.
1. Install the dummy service provider
    ```
    kore apply -f examples/service-provider-dummy.yml
    ```
1. Go through the UI and test all service-related functionality.
1. Disable `application_services` on the UI and check again.
1. Disable `services` on the UI and check again.

Functionality to check:
 * http://localhost:3000/configure/services - enable/disable services, manage services
 * create/update/delete a cloud service
 * create/update/delete an application service
 * create/update/delete service access
 * create a cloud service using the Kore CLI
    ```
    $ kore create service cloud-test --plan dummy-aws-default --cluster kore -t kore-admin
    ```
 * create an app service using the kore CLI
    ```
    kore create service app-test --plan dummy-kubernetes-default --cluster kore -t kore-admin -n app-test-ns
    ```

mrsheepuk commented on ui/lib/components/services/ServiceKindList.js:
```diff
@@ -27,10 +28,14 @@ export default class ServiceKindList extends React.Component {
     let kinds = await (await KoreApi.client()).ListServiceKinds()
-    // TODO: use a more general way of hiding the kore internal service kind "app", using a label or annotation
-    kinds.items = kinds.items.filter(k => k.metadata.name !== 'app')
+    kinds.items = kinds.items.filter(k => !k.metadata.annotations || k.metadata.annotations['kore.appvia.io/system'] !== 'true')
```
ideally (doesn't have to be part of this PR) we should add a CRD helper for the 'system' check like we have for the 'read-only' check.

bandesz replied:
I've got tired of this PR, so I'm going to skip this :(

mrsheepuk replied:
np

mrsheepuk commented:
Reviewed the Go side, all looks good to me - @davet1985 is testing so happy for him to approve once he's happy.

davet1985 commented:
**Backward-compatibility issue:**

I'm still able to delete the `flux-helm-operator` and `kube-app-manager` "system" services on existing clusters. I'm assuming the annotation is not present.

**Other issues:**

On team page clusters tab, when multiple clusters exist all namespaces are appearing under each cluster, see screenshot. (this bug might already exist)

<img width="620" alt="Screen Shot 2020-06-11 at 10 04 12" src="https://user-images.githubusercontent.com/1334068/84371920-49ed6d00-abd2-11ea-92dc-7373ac42d2ed.png">

When adding a cloud service, the option to add service access to existing cluster namespaces is still showing even if the ServiceKind has `ServiceAccessEnabled=false`, if attempted it will fail with the following
```
fieldErrors: [
  {
    field: 'kind',
    errCode: 'notAllowed',
    message: '"sqs" does not support service access'
  }
]
```

When creating a new application service, the service type dropdown is auto-selecting an option, even though there are multiple in the list. It should only auto-select when there is only one available option.

When creating a new application service and opting for a new namespace, after saving the namespaces tab is not updated with the new namespace count. It takes a page refresh for the new namespace claim to appear.

After deleting all service access from the namespaces tab, I'm unable to delete the cloud service as it states there is still service access to be deleted.
After creating a cloud service, switching to the namespaces tab and creating a service access, I'm able to switch back to the cloud services tab and delete the service without it warning about the access being present. It doesn't get deleted as the request is blocked by the API.

For these two issues, the cloud services tab state probably needs to be updated to reflect the changes in the service access if it's been amended on the namespaces tab, this wasn't an issue previously because navigation between the resources required visiting separate pages/components.

davet1985 commented on ui/lib/components/layout/SiderMenu.js:
```diff
@@ -50,7 +50,7 @@ class SiderMenu extends React.Component {
       >
         {menuItem({ key: 'configure_cloud', text: 'Cloud', href: '/configure/cloud/[[...cloud]]', link: '/configure/cloud', icon: 'cloud' })}
-        {featureEnabled(KoreFeatures.APPLICATION_SERVICES) ? null :
+        {!featureEnabled(KoreFeatures.SERVICES) ? null :
```
I think this should be changed back to showing only when app services are enabled. This is because we can manage the cloud services via the Configure/Cloud page.

bandesz replied:
I don't agree with this logic, but reverted.

davet1985 commented on ui/pages/configure/services/index.js:
```diff
@@ -28,15 +29,16 @@ export default class ServiceIndexPage extends React.Component {
           <Form.Item labelAlign="left" labelCol={{ span: 4 }} label={<Text strong>Filter by platform</Text>} style={{ marginBottom: 0 }}>
-            {this.checkboxFilter('AWS')}{this.checkboxFilter('Kubernetes')}
+            {this.checkboxFilter('AWS')}
+            {featureEnabled(KoreFeatures.APPLICATION_SERVICES) ? this.checkboxFilter('Kubernetes') : null}
```
This could be removed if the Configure/Services link was only shown when app services are enabled.

bandesz replied:
Removed

bandesz commented:
Great testing!

> I'm still able to delete the `flux-helm-operator` and `kube-app-manager` "system" services on existing clusters. I'm assuming the annotation is not present.

This is caused by a "bug" in the cluster controller: https://github.com/appvia/kore/issues/945

> On team page clusters tab, when multiple clusters exist all namespaces are appearing under each cluster, see screenshot. (this bug might already exist)

Pre-existing issue, fixed.

> When adding a cloud service, the option to add service access to existing cluster namespaces is still showing even if the ServiceKind has `ServiceAccessEnabled=false`, if attempted it will fail with the following

Fixed. I've created a bunch of dummy service kinds and plans to test this. 

> When creating a new application service, the service type dropdown is auto-selecting an option, even though there are multiple in the list. It should only auto-select when there is only one available option.

Fixed.

> When creating a new application service and opting for a new namespace, after saving the namespaces tab is not updated with the new namespace count. It takes a page refresh for the new namespace claim to appear.

For this to work correctly we have to create a namespaceclaim in the API and not in the controller. https://github.com/appvia/kore/issues/944 

> After deleting all service access from the namespaces tab, I'm unable to delete the cloud service as it states there is still service access to be deleted.
> After creating a cloud service, switching to the namespaces tab and creating a service access, I'm able to switch back to the cloud services tab and delete the service without it warning about the access being present. It doesn't get deleted as the request is blocked by the API.

IMO we shouldn't rely on such a logic in the UI (as only the API should be the source of thruth) and should simply rely on an error message from the API.
If we want to enrich the error message, let's return the extra information from the API. But do not duplicate the logic from the API. - https://github.com/appvia/kore/issues/948

davet1985 commented on ui/lib/components/teams/service/ServiceBuildForm.js:
```diff
@@ -53,13 +54,9 @@ class ServiceBuildForm extends React.Component {
-      const cluster = clusters.items.find(c => c.metadata.name === ns.spec.cluster.name)
-      // In case a namespace claim exists without a cluster, no-op if no cluster exists:
-      if (!cluster) {
-        return
-      }
+    namespaceClaims.items.filter(nc => nc.spec.cluster.name === cluster.metadata.name).forEach((ns) => {
```
@bandesz @mrsheepuk did this get missed on a rebase maybe? I think the check for a cluster should remain? See https://github.com/appvia/kore/pull/934/files

mrsheepuk replied:
looks OK to me - the cluster is being passed in as a property now instead of a list of all clusters, so as cluster is a required property, we don't need to check for it not existing any more.

bandesz replied:
@davet1985 what Mark said. For now you can only create service credentials in the cluster which owns the service. (I tried to remove as little code as possible, so we don't have to work too much if we add the sharing functionality)

bandesz replied:
Also I added this change at the very last, as I just found this on my second round of testing.
